### PR TITLE
WIP: Add partial support for zwp_text_input_unstable_v3

### PR DESCRIPTION
**Pull request by dcz-purism**
Present features: preediting string, sending enter/leave events and handling basic events.

There is a rootston version taking advantage of this change to paste some text into text fields [here](https://code.puri.sm/dorota.czaplejewicz/wlroots/src/text_input_test). GTK version updated to support this input method is [here](https://code.puri.sm/dorota.czaplejewicz/gtk).

### CODE REVIEW

File: include/wlr/types/wlr_text_input.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (C) 2018 Purism SPC
+ *
+ * SPDX-License-Identifier: MIT
+ */
```
Comment (SirCmpwn) [inline include/wlr/types/wlr_text_input.h]: hm

File: include/wlr/types/wlr_text_input.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+#ifndef WLR_TYPES_WLR_GTK_TEXT_INPUT_H
+#define WLR_TYPES_WLR_GTK_TEXT_INPUT_H
```
[the remark this answers is not in the document]
Comment (SirCmpwn) [reply include/wlr/types/wlr_text_input.h]: Also this all should be s/gtk/zwp/g right

File: include/wlr/types/wlr_text_input.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	struct wlr_text_input_manager *manager);
+void wlr_text_input_send_enter(struct wlr_text_input *text_input,
+	struct wlr_surface *wlr_surface);
+void wlr_text_input_send_leave(struct wlr_text_input *text_input,
+	struct wlr_surface *wlr_surface);
+void wlr_text_input_send_preedit_string(struct wlr_text_input *text_input,
+	const char *text, uint32_t cursor);
+void wlr_text_input_send_commit_string(struct wlr_text_input *text_input,
+	const char *text);
+#endif
```
Comment (SirCmpwn) [inline include/wlr/types/wlr_text_input.h]: Add newline before endif

File: protocol/text-input-unstable-v3.xml [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+    SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
+    AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
+    ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF
+    THIS SOFTWARE.
+  </copyright>
+
+  <interface name="zwp_text_input_v3" version="1">
+    <description summary="text input">
+      The zwp_text_input_v3 interface represents text input and input methods
```
Comment (SirCmpwn) [inline protocol/text-input-unstable-v3.xml]: I thought this was on v2?

File: include/wlr/types/wlr_text_input.h (hunk 1 above)
Comment (dcz-purism) [reply include/wlr/types/wlr_text_input.h]: Please correct me if the license is wrong - I tried to match the identifier to the LICENSE file.

**Comment by SirCmpwn** [comment]
I like where this is going

**Comment by dcz-purism** [comment]
Changes implemented in this revision:

- removed header comments
- removed `manager::text_input`
- sorted `include`s
- fixed name of include guard to reflect file name
- added newline before `endif`

Not changed yet:

- serial number checking, since its purpose is unclear (waiting for feedback)

**Comment by SirCmpwn** [comment]
>serial number checking, since its purpose is unclear (waiting for feedback)

We can probably just yank this and mention it when we send the protocol around for review.

File: include/wlr/types/wlr_text_input.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
+#ifndef WLR_TYPES_WLR_TEXT_INPUT_H
+#define WLR_TYPES_WLR_TEXT_INPUT_H
+
+#include <wayland-server.h>
+#include <wlr/types/wlr_surface.h>
+
+
```
Comment (SirCmpwn) [inline include/wlr/types/wlr_text_input.h]: Spacing nits: no newline between `#define` and `#include`s, and only one space between `#include`s and declarations

File: include/wlr/types/wlr_text_input.h [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+struct wlr_text_input {
+	struct wl_resource *resource;
+	struct wlr_surface *focused_surface;
+	uint32_t next_serial;
+	struct {
+		struct wl_signal enable_notify;
+		struct wl_signal disable_notify;
```
Comment (SirCmpwn) [inline include/wlr/types/wlr_text_input.h]: `s/_notify//`

File: types/wlr_text_input.c [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+#include <stdlib.h>
+#include <stddef.h>
+
+#include <wlr/types/wlr_seat.h>
+#include <wlr/types/wlr_text_input.h>
+
+#include <wlr/util/log.h>
+
+#include "text-input-unstable-v3-protocol.h"
+#include "util/signal.h"
```
Comment (SirCmpwn) [inline types/wlr_text_input.c]: More spacing nit: remove all of the empty lines above here and one of the empty lines below here

File: types/wlr_text_input.c [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+void wlr_text_input_send_commit_string(struct wlr_text_input *text_input,
+		const char *text) {
+	zwp_text_input_v3_send_commit_string(text_input->resource, text);
+}
+
+static void text_input_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wlr_log(L_INFO, "STUB: text_input_destroy");
```
Comment (SirCmpwn) [inline types/wlr_text_input.c]: Can you implement this

File: types/wlr_text_input.c [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+static void text_input_enable(struct wl_client *client,
+		struct wl_resource *resource, uint32_t serial,
+		uint32_t show_input_panel) {
+	struct wlr_text_input *text_input = text_input_from_resource(resource);
+	if (serial >= text_input->next_serial) {
+		wlr_log(L_INFO,
+			"text_input_enable called with a serial that wasn't sent yet: %d",
+			serial);
```
[the remark this answers is not in the document]
Comment (SirCmpwn) [reply types/wlr_text_input.c]: bump

File: types/wlr_text_input.c [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+static void text_input_set_cursor_rectangle(struct wl_client *client,
+		struct wl_resource *resource, int32_t x, int32_t y, int32_t width,
+		int32_t height) {
+	wlr_log(L_INFO, "STUB: set_cursor_rectangle(%d, %d, %d, %d)", x, y, width,
+		height);
+}
+
+static void text_input_commit(struct wl_client *client,
+		struct wl_resource *resource) {
+	wlr_log(L_INFO, "STUB: text_input_commit()");
```
Comment (SirCmpwn) [inline types/wlr_text_input.c]: Please implement

File: types/wlr_text_input.c [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+static void text_input_set_content_type(struct wl_client *client,
+		struct wl_resource *resource, uint32_t hint, uint32_t purpose) {
+	wlr_log(L_INFO, "STUB: set_content_type(%d, %d)", hint, purpose);
+}
+
+static void text_input_set_cursor_rectangle(struct wl_client *client,
+		struct wl_resource *resource, int32_t x, int32_t y, int32_t width,
+		int32_t height) {
+	wlr_log(L_INFO, "STUB: set_cursor_rectangle(%d, %d, %d, %d)", x, y, width,
+		height);
```
Comment (SirCmpwn) [inline types/wlr_text_input.c]: This stuff too, you can just stash it and not use it on rootston but definitely implement the state transitions

File: types/wlr_text_input.c [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+	.set_content_type = text_input_set_content_type,
+	.set_cursor_rectangle = text_input_set_cursor_rectangle,
+	.commit = text_input_commit,
+};
+
+static const struct zwp_text_input_manager_v3_interface text_input_manager_impl;
+
+static void text_input_manager_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wlr_log(L_INFO, "STUB: text_input_manager_destroy");
```
Comment (SirCmpwn) [inline types/wlr_text_input.c]: Implement this dude too

Basically this file shouldn't have stubs, even if the rootston implementation doesn't take advantage of everything we can still provide a complete text_input state machine implementation reasonably easily here

**Comment by dcz-purism** [comment]
Changes:
- yanked serial and replaced it with `0` in calls
- implemented stubs by exposing them as signals
- whitespace fixes
- removed `_notify` from signal names

File: include/wlr/types/wlr_text_input.h [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+	struct {
+		int32_t x;
+		int32_t y;
+		int32_t width;
+		int32_t height;
+	} cursor_rectangle;
+};
+
+struct wlr_text_input {
+	struct wlr_seat *seat; // used only for unlinking seat->text_input; perhaps better to use the destroy signal
```
Comment (SirCmpwn) [inline include/wlr/types/wlr_text_input.h]: This sort of comment is better put into the commit message or here on GitHub

**Comment by SirCmpwn** [comment]
>yanked serial and replaced it with 0 in calls

I was expecting you to remove it from the protocol

>implemented stubs by exposing them as signals

Hmm, this doesn't look like I expected. It looks like you're throwing away the state when it's committed? That doesn't make sense. You should check out the commit flow that the shells use, the purpose is to double buffer the state. Should also update the xml descriptions to use the usual language around double buffering that we see in other protocols.

**Comment by emersion** [comment]
>I was expecting you to remove it from the protocol

Ugh. What about posting in wayland-devel and keeping it while waiting for a protocol v2?

File: include/wlr/types/wlr_text_input.h (hunk 13 above)
Comment (dcz-purism) [reply include/wlr/types/wlr_text_input.h]: Sorry, forgot to move it before committing. What do you think about the idea in the comment anyway?

**Comment by SirCmpwn** [comment]
No, this is part of the process. No one is going to push this protocol forward for us. Take the initiative to sponsor v3 by cleaning it up, iterating it through us and other stakeholders (KDE, GTK) for feedback, then posting your version to wayland-devel for consideration. As part of that process it's useful to have an example implementation of the protocol to show off - that's where this PR comes in.

**Comment by dcz-purism** [comment]
@emersion: my email to wayland-devel is stuck in moderation...

File: include/wlr/types/wlr_text_input.h (hunk 13 above)
Comment (SirCmpwn) [reply include/wlr/types/wlr_text_input.h]: Probably better to use the destroy signal, yeah.

Comment (emersion) [reply include/wlr/types/wlr_text_input.h]: I don't think so. That would require adding text_input code in the seat.

Comment (SirCmpwn) [reply include/wlr/types/wlr_text_input.h]: Ah, good point. This patch already touches the seat, should probably be cleaned out.

**Comment by dcz-purism** [comment]
About throwing away state: initially I implemented it with double-buffering, before realizing that it doesn't make sense to keep `set_surrounding_text` state between calls. No one wants to paste the same text after the content type has changed. I'm under the impression that the `commit` is only meant to convey what was changed from the most recent `commit` call. Perhaps `set_content_type` and `set_cursor_rectangle` might be kept between calls, but I also don't necessarily see them as state.

**Comment by SirCmpwn** [comment]
>Perhaps set_content_type and set_cursor_rectangle might be kept between calls, but I also don't necessarily see them as state.

I mean, they are

File: include/wlr/types/wlr_text_input.h (hunk 13 above)
Comment (dcz-purism) [reply include/wlr/types/wlr_text_input.h]: Seat code is touched because there needs to be some way to get notified of new `text_input` instances. Since only one per seat is allowed, I thought seat is [the best place](https://github.com/swaywm/wlroots/pull/776/files#diff-6eb821df1a7fff48e265e52e0a742741R215). Should it be placed somewhere else instead?

Comment (SirCmpwn) [reply include/wlr/types/wlr_text_input.h]: Probably better to construct a text_input with a seat reference and have it manage it. Or let the compositor glue stuff together.

**Comment by dcz-purism** [comment]
Changes:
- removed new code from `wlr_seat`
- removed serial from protocol
- implemented double buffering of state
- abandoned clearing of state between commits; this means client needs to keep track of what's in `set_surrounding_text`
- moved create signal to manager

File: include/wlr/types/wlr_text_input.h [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+#include <wayland-server.h>
+#include <wlr/types/wlr_seat.h>
+#include <wlr/types/wlr_surface.h>
+
+struct wlr_text_input_atomic_properties {
+	struct {
+		char *text;
+		uint32_t cursor;
+		uint32_t anchor;
+	} surrounding;
```
Comment (SirCmpwn) [inline include/wlr/types/wlr_text_input.h]: Nitpick, add a newline after this

File: include/wlr/types/wlr_text_input.h [hunk 15]
```diff
@@ -0,0 +1,67 @@
+#ifndef WLR_TYPES_WLR_TEXT_INPUT_H
+#define WLR_TYPES_WLR_TEXT_INPUT_H
+
+#include <wayland-server.h>
+#include <wlr/types/wlr_seat.h>
+#include <wlr/types/wlr_surface.h>
+
+struct wlr_text_input_atomic_properties {
```
Comment (SirCmpwn) [inline include/wlr/types/wlr_text_input.h]: Nitpick, just call this wlr_text_input_properties

File: include/wlr/types/wlr_text_input.h [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+		int32_t height;
+	} cursor_rectangle;
+};
+
+struct wlr_text_input {
+	struct wlr_seat *seat; // used only for unlinking seat->text_input; perhaps better to use the destroy signal
+	struct wl_resource *resource;
+	struct wlr_surface *focused_surface;
+	struct wlr_text_input_atomic_properties pending_properties;
+	struct wlr_text_input_atomic_properties properties;
```
Comment (SirCmpwn) [inline include/wlr/types/wlr_text_input.h]: We use `pending` and `current` as field names for this sort of thing.

**Comment by SirCmpwn** [comment]
This is getting pretty close, I like it.

1. The XML descriptions should use the conventional language for describing double buffered state (see e.g. wl_surface, xdg_surface, etc)
2. Is there a test client we can use?

**Comment by SirCmpwn** [comment]
>abandoned clearing of state between commits; this means client needs to keep track of what's in set_surrounding_text

Can you elaborate on this?

**Comment by dcz-purism** [comment]
Updated noticed nitpicks.

**Comment by dcz-purism** [comment]
1. I'll try to do that.
2. There is a test client at https://code.puri.sm/dorota.czaplejewicz/wlroots/src/text_input_test , although it needs to be updated after seat code was cleaned (the newest update was not tested outside of compilation). It tests only the basic preedit/commit_string functionality. See my original comment on this merge request.

I'm not 100% sure how `set_surrounding_text` is meant to work, but assuming the client almost never wants to set the same non-null text twice, it can no longer assume that the text will be empty after each commit.

**Comment by SirCmpwn** [comment]
Well the text wouldn't be empty after commit, would it be? Why would it be?

**Comment by SirCmpwn** [comment]
>needs to be updated

Please do that and give us a test plan so we can make sure this works

**Comment by dcz-purism** [comment]
What form would you like the test plan to take?

I'm not sure what other implementations will do when a commit is called; if they come to the conclusion (as I did) that surrounding is invalid between contexts, and thus commit calls, they could clear it.

Either way, it's not cleared now.

**Comment by SirCmpwn** [comment]
>What form would you like the test plan to take?

Just a list of steps we can perform to exercise this code and verify it works as promised

>I'm not sure what other implementations will do when a commit is called; if they come to the conclusion (as I did) that surrounding text is not valid across contexts, and thus commit calls, they could clear it.

If you cleared it out after commit, then when is the server ever going to be able to display it?

**Comment by dcz-purism** [comment]
The scenario in my mind is this:

- text field A gets focus
- client sends content_purpose = email
- client sends set_surrounding_text = "@foobar.com"
- client commits
- field B gets focused
- client uses the same text_input to set content_purpose = normal
- client commits

If commit doesn't clean up set_surrounding_text, then the client must ensure it's correct for the given "context". In this case, field B would get "@foobar.com" as the prompt.

If the compositor is made under the assumption that set_surrounding_text is ephemeral and clear by default, clients will not get accidental leftovers when switching fields.

Perhaps this is underspecified; or perhaps this should be prevented by replacing the separate calls with a single compound one to reduce the potential for such bugs. I don't know how much benefit there is from keeping those calls separate though.

**Comment by SirCmpwn** [comment]
I see the sense in your argument, but it works a bit differently from how most wayland protocols are assumed to work. I'm okay with your approach if you document it clearly in the XML.

**Comment by dcz-purism** [comment]
I can think of a couple of approaches. I would try to solve it first by changing the protocol in such a way that making an error is not possible first. In order of preference:
- turn 3 calls and commit into a single call: the code generator/compiler enforces compliance
- clear state on each focus change (`enable` request or `enter` event): this is closest to my intuitive understanding of when state changes, aligning behaviour with expectation
- clear state on each commit: this will force clients to re-request all state each time even within a single field, in the likely event of forgetting this will result in bugs (resetting field to default)
- do not clear state on each commit: similar as above, any bugs will be more confusing because they will come from old values
- only reset surrounding text: splitting state into permanent and volatile is even more surprising IMO

If there's a way to enforce compliance automatically and keep separate calls for each property, I'm all ears.

**Comment by SirCmpwn** [comment]
I don't want to merge requests, I think that the several requests plus double buffering approach works well (and is better for future extensibility, we can add more double buffered requests later more easily).

>clear state on each commit: this will force clients to re-request all state each time even within a single field, in the likely event of forgetting this will result in bugs (resetting field to default)

I like this. We can make not setting the text before commit a protocol error.

**Comment by dcz-purism** [comment]
> We can make not setting the text before commit a protocol error.
Singling out just the text property for special treatment is a bug invitation IMO, and future implementations would break compatibility if they tried to introduce another such property.
If state should change on commit and not focus switch, I'd make it reset to a set of predefined safe defaults.

**Comment by SirCmpwn** [comment]
>Singling out just the text property for special treatment is a bug invitation

I don't think this would lead to bugs

>future implementations would break compatibility if they tried to introduce another such property

This would naturally necessitate a version bump and the compositor can handle it accordingly, wayland is designed to handle such cases.

**Comment by dcz-purism** [comment]
What criterion would you propose to separate properties resetting on commit and those which should stay the same?

**Comment by dcz-purism** [comment]
Updated:
- double-buffering language in protocol description
- described initial values
- minor bugs

**Comment by SirCmpwn** [comment]
>What criterion would you propose to separate properties resetting on commit and those which should stay the same?

Discussion on GitHub/MLs?

File: protocol/text-input-unstable-v3.xml [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+      <description summary="set content purpose and hint">
+       Sets the content purpose and content hint. While the purpose is the
+       basic purpose of an input field, the hint flags allow to modify some
+       of the behavior.
+       
+       Content type is double-buffered state, see zwp_text_input_v3.commit.
+
+       zwp_text_input_v3.commit copies the pending content type to the
+       current content type. Otherwise the pending and current content types
+       never change.
```
Comment (SirCmpwn) [inline protocol/text-input-unstable-v3.xml]: This language still doesn't quite match. Try:

>Values set in this way are double-buffered. They will get applied on the next commit.

Also, don't duplicate docs on what commit does in every request.

Comment (dcz-purism) [reply protocol/text-input-unstable-v3.xml]: Updated - I'm not sure if the update is what you're looking for; the text was copied almost verbose from wayland.xml.

Comment (SirCmpwn) [reply protocol/text-input-unstable-v3.xml]: Ah, I was looking at xdg-shell. I guess the "standard langauge" is a bit nebulous. xdg-shell's language is a big more standard now because I copied it for layer-shell and now you've copied it for text input :tada:

**Comment by SirCmpwn** [comment]
Can you send the current iteration of the protocol around to the main stakeholders for comments?

Also still looking for a way to test this, but there's no rush, take your time.

**Comment by dcz-purism** [comment]
Current iteration of the protocol (can't be attached here)

https://paste.debian.net/1018117/

**Comment by dcz-purism** [comment]
WIP testing procedure: use `text_input_test` branch of rootston from [here](https://code.puri.sm/dorota.czaplejewicz/wlroots/src/text_input_test).

Only 1 client at a time supported. Works with gtk 3.22's wayland IM module from https://code.puri.sm/dorota.czaplejewicz/gtk .

To test with older GTK installed, build the altered version and make use only of the loadable input method module with a program of your choice (`yad` here):

```
$ chrpath -d .../gtk-install/lib/gtk-3.0/3.0.0/immodules/im-wayland.so
$ rootston
$ WAYLAND_DEBUG=1 GDK_BACKEND=wayland GTK_IM_MODULE=wayland GTK_IM_MODULE_FILE=.../gtk-install/lib/gtk-3.0/3.0.0/immodules.cache yad --entry
```

The procedure is unlikely to change with subsequent updates.

**Comment by SirCmpwn** [comment]
I was able to get this sort of... working? I'm not really sure. What is the expected behavior? Can you elaborate on your test plan?

**Comment by dcz-purism** [comment]
Sorry - I should have added a bit.

The demo works by sending the `enter`, `leave`, `preedit_string` and `commit_string` requests. It will handle the creation of a new text input. The `enter` and `leave` are sent automatically, the `preedit_string` and `commit_string` requests are sent when a field is in focus and the mouse is left-clicked. The first 2 events are `preedit_string`, and the last one is `commit_string`, all with predefined strings "str2", "string1", "textfinal". After `commit_string` is submitted, the instance of `text_input` will not send any more string events.

Events attached to the `text_input` instance are not handled. The request `delete_surrounding_text` is not sent.

I'm not sure how the GTK side should be interpreting this - as far as the protocol goes, the messages are being exchanged correctly, but `commit_string` doesn't seem to be clearing the previously received preedit strings, which may be a bug with original GTK implementation.

**Comment by SirCmpwn** [comment]
I'm still not entirely clear on what buttons I need to press in which order and what I should see when I know it's working.

**Comment by dcz-purism** [comment]
1. Run a GTK program,
2. Get a text field focused
3. Left-click anywhere, observe text suggestion
4. Left-click anywhere, observe different text suggestion
5. Left-click anywhere, observe committed text
6. Back to normal

**Comment by SirCmpwn** [comment]
Okay, I was able to confirm that works.

**Comment by SirCmpwn** [comment]
I think this is only a small piece of a rather broad pie, though. I'm not sure what the future direction is and how we should incoroproate all of it (and in what degrees at what times) into wlroots.

**Comment by dcz-purism** [comment]
The protocol support is rather self-contained.
There needs to be something to handle it on the compositor side, of course. There are two basic possibilities are to make an input method part of the compositor or not.

In the former case, this protocol already gives enough support on the wlroots side to make this a reality. For interfacing with a separate input method, there's input-method-unstable-v1, which can also make its way into wlroots. That was in fact my next planned step after this.

To summarize, I think the text-input protocol on its own is already useful, and it's going to be even more useful together with input-method while not interfering with anything else in wlroots if compositor authors decide not to make use of them.

**Comment by SirCmpwn** [comment]
Alright, thanks for clarifying.

**Comment by Shugyousha** [comment]
This WIP looks very interesting! The protocol suggestion (with links to the implementation) should be sent to the wayland mailing list soon.

Currently there was a different version 3 of the text-input protocol posted there: https://patchwork.freedesktop.org/patch/209838/ some of the comments garnered there could be applied here as well.

**Comment by dcz-purism** [comment]
I've updated the XML file with fixes indicated in the wayland-devel thread.

**Comment by SirCmpwn** [comment]
Can you also reply to the wayland-devel thread updating them about the work we've been doing over here? Be sure to reply-all so the Gnome guys gets it in his inbox.

**Comment by dcz-purism** [comment]
Sure. In the meantime, I debugged multi-client support in the demo and pushed it [here](https://code.puri.sm/dorota.czaplejewicz/wlroots/src/text_input_test). Hopefully it doesn't crash too much.

**Comment by SirCmpwn** [comment]
I would love to see your Rootston changes added to this patch, and perhaps a test client in examples/ if you have time.

**Comment by dcz-purism** [comment]
Are you sure about Rootston changes? The "click to have arbitrary text appear" behaviour is going to make rootston rather useless for clients supporting this protocol.
I could see "noop" support added to rootston: all the structs and signals in the right places.

Anyway, I'm beginning work on `zwp_input_method`, which I think should give enough basis for rootston support without reimplementing a virtual keyboard inside rootston.

**Comment by SirCmpwn** [comment]
Oh, alright then.

**Comment by dcz-purism** [comment]
Some highlights from the IRC discussion about text input.

The current protocol does not support highlighting a part of preedit text in order to indicate that it's focus (important for Japanese, Korean, Chinese input methods). Ibus protocol solves this by passing around styling information between the input method and the client. Possible appraoaches:

1. Communicate information about what is active and leave styling to application. Possibly also about what is wrong, or what's in rubies etc. Con: limited possibilities.
2. Communicate styling directly via some form of rich text. Con: style like colors will probably clash with the application theme.
3. Leave styling to input method using an additional surface. Con: can never match up styling with the application.

Option 3 is not mutually exclusive with the others.

The consensus seemed to be to choose a version of option 1, possibly limiting it to the bare minimum: 1 possible currently active range. This could be related to the cursor position which is already present in `set_preedit_string`.
